lib/mspec/runner/actions/constants_leak_checker.rb

```diff
@@ -0,0 +1,86 @@
+class ConstantsLeakChecker
+  attr_accessor :constants_new
+
+  def initialize
+    @constants_before = Module.constants
+  end
+
+  def check
+    @constants_after = Module.constants
+    @constants_new = @constants_after - @constants_before
+    @constants_new.empty?
+  end
+end
+
+class ConstantsLockFile
+  LOCK_FILE_NAME = '.mspec.constants.lock'
+
+  def self.exist?
+    File.exist?(LOCK_FILE_NAME)
+  end
+
+  def self.load
+    File.read(LOCK_FILE_NAME).split
+  end
+
+  def self.dump(ary)
+    File.write(LOCK_FILE_NAME, ary.map(&:to_s).uniq.sort.join("\n"))
+  end
+end
+
+class ConstantsLeakLockAction
+  def register
+    MSpec.register :start, self
+    MSpec.register :finish, self
+  end
+
+  def start
+    @checker = ConstantsLeakChecker.new
+  end
+
+  def finish
+    unless @checker.check
+      if ConstantsLockFile.exist?
+        constants_locked = ConstantsLockFile.load
+        result = constants_locked | @checker.constants_new.map(&:to_s)
+        ConstantsLockFile.dump(result)
+      else
+        ConstantsLockFile.dump(@checker.constants_new)
+      end
+    end
+  end
+end
+
+class ConstantsLeakCheckerAction
```

Comment:
I would use a simple action for both checking & recording, since they are very similar.